pkg/k8sops/generators.go

```diff
@@ -0,0 +1,392 @@
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// GenerateCRD generates the crd object needed for the M3DBCluster
+func (k *K8sops) GenerateCRD() *apiextensionsv1beta1.CustomResourceDefinition {
+	return &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: m3dboperator.Name,
+		},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+			Group:   m3dboperator.GroupName,
+			Version: m3dboperator.Version,
+			Scope:   apiextensionsv1beta1.NamespaceScoped,
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural: m3dboperator.ResourcePlural,
+				Kind:   m3dboperator.ResourceKind,
+			},
+		},
+	}
+}
+
+// GenerateService generates the Service object for m3coordinator
+func (k *K8sops) GenerateService(name string) (*v1.Service, error) {
+
+	switch name {
+	case M3dbNodeSvcName:
+		ports, err := k.GenerateServicePorts(name)
+		if err != nil {
+			return nil, err
+		}
+		return &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: M3dbNodeSvcName,
+				Labels: map[string]string{
+					"app": M3dbNodeSvcName,
+				},
+			},
+			Spec: v1.ServiceSpec{
+				Selector: map[string]string{
+					"app": M3dbNodeSvcName,
+				},
+				Ports:     ports,
+				ClusterIP: "None",
+			},
+		}, nil
+	case M3coordinatorSvcName:
+		ports, err := k.GenerateServicePorts(name)
+		if err != nil {
+			return nil, err
+		}
+		return &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: M3coordinatorSvcName,
+				Labels: map[string]string{
+					"app": M3coordinatorSvcName,
+				},
+			},
+			Spec: v1.ServiceSpec{
+				Selector: map[string]string{
+					"app": M3dbNodeSvcName,
+				},
+				Ports: ports,
+			},
+		}, nil
+	default:
+		return nil, errors.New("service is not defined")
+	}
+}
+
+// GenerateServicePorts generates the service ports object for m3db
+func (k *K8sops) GenerateServicePorts(name string) ([]v1.ServicePort, error) {
+	switch name {
+	case M3dbNodeSvcName:
+		return []v1.ServicePort{
+			v1.ServicePort{
+				Name:     "client",
+				Port:     9000,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "cluster",
+				Port:     9001,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "http-node",
+				Port:     9002,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "http-cluster",
+				Port:     9003,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "debug",
+				Port:     9004,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "query",
+				Port:     7201,
+				Protocol: v1.ProtocolTCP,
+			},
+			v1.ServicePort{
+				Name:     "query-metrics",
+				Port:     7203,
+				Protocol: v1.ProtocolTCP,
+			},
+		}, nil
+	case M3coordinatorSvcName:
+		return []v1.ServicePort{
+			v1.ServicePort{
+				Name:     "query",
+				Port:     7201,
```

Comment:
nit: should probably expose configs for all the ports to allow users to change em

Reply:
Done, but see comment above